src/scheduler/scheduler.service.ts

```diff
@@ -0,0 +1,83 @@
+      if (schedule.recursion && schedule.recursion.days) {
+        //handle start day doesn't match recurring pattern
+        const cronExpArray = this.getCustomCronExpression(schedule.recursion, schedule.startDate).split(' ');
+        const IsStartDayMatchCron = this.toCsvString(schedule.recursion.days) === cronExpArray[cronExpArray.length - 1];
+        if (IsStartDayMatchCron) callback();
+      } else callback();
+      job.start();
+    });
+  }
+  scheduleJob(jobName: string, startDate: Date, callback: () => void) {
+    if (moment(startDate).utcOffset(0).milliseconds() < Date.now()) throw new RangeError('Invalid Date: date should be after present');
+    const name = `${jobName}-timeout`;
+    const timeDiff = moment(startDate).utcOffset(0).diff(moment(), 'milliseconds');
+    this.addTimeout(name, timeDiff, callback);
+  }
+  private addTimeout(name: string, ms: number, callback: () => void) {
+    const timeout = setTimeout(callback, ms);
+    if (this.schedulerRegistry.getTimeouts().find((timeoutName) => timeoutName === name)) this.schedulerRegistry.deleteTimeout(name);
+    this.schedulerRegistry.addTimeout(name, timeout);
+  }
+  private addRecurringJob({ name, predefined, recursion, startDate }: Schedule, callback: () => void) {
+    const jobId = `${name}-recurring`;
+    if (this.schedulerRegistry.getCronJobs().has(jobId)) this.schedulerRegistry.deleteCronJob(jobId);
+    const cronExp = predefined ? this.getPredefinedCronExpression(predefined, startDate) : this.getCustomCronExpression(recursion, startDate);
+    const job = new CronJob(cronExp, callback);
+    this.schedulerRegistry.addCronJob(jobId, job);
+    return job;
+  }
+  private getPredefinedCronExpression(interval: RecurringInterval, date: Date): string {
+    const dateUtc = moment(date).utcOffset(0).toDate();
+    let exp = '';
+    switch (interval) {
+      case RecurringInterval.EVERY_DAY:
+        exp = `0 ${dateUtc.getMinutes()} ${dateUtc.getHours()} * * *`;
+        break;
+      case RecurringInterval.EVERY_MONTH:
+        exp = `0 ${dateUtc.getMinutes()} ${dateUtc.getHours()} ${dateUtc.getDate()} * *`;
+        break;
+      case RecurringInterval.EVERY_WEEK:
+        exp = `0 ${dateUtc.getMinutes()} ${dateUtc.getHours()} * * ${dateUtc.getDay()}`;
+        break;
+      case RecurringInterval.EVERY_10_MINUTES:
+        exp = CronExpression.EVERY_10_MINUTES;
+      default:
+        throw new Error('Invalid Enum value: RecurringInterval');
+    }
+    return exp;
+  }
+  private toCsvString(arr: any[]) {
+    const tempArr = arr.map((v) => v.toString());
+    return tempArr.reduce((acc, curr) => `${acc},${curr}`);
+  }
+  private getCustomCronExpression(custom: Recursion, date: Date): string {
+    const dateUtc = moment(date).utcOffset(0).toDate();
+    let exp = '';
+    switch (custom.period) {
+      case RecursionPeriod.MONTH:
+        const dateRange = custom.dates ? this.toCsvString(custom.dates) : date.getDate();
```

Comment:
Since u use dateRange, what if the recurring is every 5, 10, 16 of month?

Reply:
Sorry for the bad naming. Date range here means list of dates in comma-separated string, like "5,10,16". I'll rename this